File: src/api/discussion-send-draft.ts
import { Discussion, DiscussionDraftMessage, DiscussionMessages, SessionHandle } from "~/models";
import { discussionPostCommand } from "./private/discussion-post-command";
import { encodeDiscussionSendAction } from "~/encoders/discussion-send-action";
import { discussions } from "./discussions";
import { discussionMessages } from "./discussion-messages";

export const discussionSendDraft = async (session: SessionHandle, discussion: Discussion, draft: DiscussionDraftMessage, includeParentsAndStudents = false): Promise<void> => {
  if (typeof discussion.messages?.sendAction === "undefined")
    throw new Error("You can't create drafts in this discussion.");

  await discussionPostCommand(session, "", {
    button: encodeDiscussionSendAction(discussion.messages.sendAction, includeParentsAndStudents),
    content: draft.content,
    id: draft.possessionID,
    replyMessageID: draft.replyMessageID
  });

  await discussions(session, discussion.cache);
  await discussionMessages(session, discussion);
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/discussion-send-draft.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.api.private.discussionPostCommand
import ink.literate.pawnote.encoders.encodeDiscussionSendAction
import ink.literate.pawnote.models.Discussion
import ink.literate.pawnote.models.DiscussionDraftMessage
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.errors.DiscussionActionError
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

suspend fun discussionSendDraft (session: SessionHandle, discussion: Discussion, draft: DiscussionDraftMessage, includeParentsAndStudents: Boolean = false) {
    if (discussion.messages == null || discussion.messages!!.sendAction == null)
        throw DiscussionActionError()

    discussionPostCommand(session, null, buildJsonObject {
        put("button", encodeDiscussionSendAction(discussion.messages!!.sendAction!!, includeParentsAndStudents).code)
        put("content", draft.content)
        put("id", draft.possessionID)
        put("replyMessageID", draft.replyMessageID)
    })

    discussions(session, discussion.cache)
    discussionMessages(session, discussion)
}